pinc/DailyPageLimit.inc

```diff
@@ -0,0 +1,132 @@
+<?php
+include_once($relPath."Project.inc");
+include_once($relPath."misc.inc"); // functions that might be used in project_selector
+
+class DailyPageLimit
+{
+    function DailyPageLimit( $name, $project_selector, $daily_limit )
+    {
+        $this->name = $name;
+        $this->project_selector = $project_selector;
+        $this->daily_limit = $daily_limit;
+    }
+
+    function applies_to_project($project)
+    {
+        $str_to_eval = "return $this->project_selector;";
+        // echo "<br>\nevaluating: $str_to_eval<br>\n";
+        $result = eval($str_to_eval);
+        // echo "returns "; var_dump($result);
+        return $result;
+    }
+
+    function get_current_count_for_user($username)
+    {
+        $res = mysqli_query(DPDatabase::get_connection(), "
+            SELECT current_count
+            FROM daily_page_limit_user_counts
+            WHERE
+                limit_name = '$this->name'
+                AND
+                username = '$username'
+        ") or die(mysqli_error(DPDatabase::get_connection()));
+        if (mysqli_num_rows($res) == 0)
+        {
+            return 0;
+        }
+        else
+        {
+            list($count) = mysqli_fetch_row($res);
+            return $count;
+        }
+    }
+
+    function user_count_has_reached_limit($username)
+    {
+        $user_count = $this->get_current_count_for_user($username);
+        return ($user_count >= $this->daily_limit);
+    }
+
+    function adjust_user_count($username, $adjustment)
+    {
+        $res = mysqli_query(DPDatabase::get_connection(), "
+            INSERT INTO daily_page_limit_user_counts
+            SET
+                limit_name = '$this->name',
+                username = '$username',
+                current_count = $adjustment
+            ON DUPLICATE KEY UPDATE
+                current_count = current_count + $adjustment
+        ") or die(mysqli_error(DPDatabase::get_connection()));
+        // $n_rows = mysqli_affected_rows($res);
+        // $n_rows == 0 means an existing row was set to its current values
+        // $n_rows == 1 means a new row was inserted
+        // $n_rows == 2 means an existing row was updated
+    }
+}
+
+// We assume that there will be very few DPLs in the system,
+// and a project will rarely if ever be targeted by more than one DPL.
+
+function dpls_get_all()
+{
+    static $all_dpls = NULL;
+    if (is_null($all_dpls))
+    {
+        // XXX get from DB
+        $all_dpls = array(
+            new DailyPageLimit('F1', '$project->state == "F1.proj_avail" && !startswith($project->nameofwork, "F1 self-evaluation")', 3)
```

Comment:
Rather than make this so extensible and requiring an eval(), why not make it narrower for the few cases that pgdp.net needs?

Reply:
If we don't represent (the details of) the daily page limits in the database, then we can certainly avoid the eval.

Reply:
We should consider putting the creation of `DailyPageLimit` objects for a site into a separate file. If the file exists we `include()` the file at the bottom of `pinc/DailyPageLimit.inc` which would then create the objects. This would allow us (and others) to have a file we can slide into place that would define these without patch hacks or putting them into the codebase.

I'm doing something similar for Glyphsets and it feels like a good approach.

Reply:
Oh certainly, if DailyPageLimits were going to be created explicitly (rather than slurped in from a database table), then I'd define them in a separate file. (That just didn't seem like an option 4 years ago.) Putting the file of declarations outside the code-base seems fine too. Would we look for it in $dyn_dir?

Reply:
Well, we could do it either like we do for `$site_registration_protection_code` (having `site_vars.php` explicitly point to the file) or have the code assume some filename within `pinc/` and if it exists, include it.

Reply:
So if the file were in `pinc`, it'd be put there via `customize_deployment` (i.e., in our case, by copying from `c_hacks.newfiles`)?

Reply:
Correct